docs/worker-versioning.md

```diff
@@ -0,0 +1,229 @@
+Worker Versioning
+=================
+
+**Note 1:** In this iteration of Worker Versioning we deprecated the Version Set concept 
+and APIs related to it. If you are using old Worker Versioning APIs please migrate
+to the new APIs using the process [outlined in a later section](#migrating-from-version-sets).
+
+**Note 2:** Worker Versioning is still considered experimental and not recommended 
+for production usage. Future breaking changes may be made in deemed necessary.
+
+Worker Versioning simplifies the process of deploying changes to [Worker Programs](https://docs.temporal.io/workers#worker-program). 
+It does this by letting you specify a Build ID for your Worker. Temporal Server
+uses the Build ID to route each Workflow and/or Activity to a Worker instance that
+can process it.
+
+Worker Versioning guarantees that Workflow Executions started on a particular Build ID
+will only be processed by Workers of the same Build ID, unless instructed otherwise 
+(via [Redirect Rules](#redirect-rules)). With this guarantee, you can make any
+([non-deterministic](https://docs.temporal.io/workflows#non-deterministic-change)) 
+change to your Worker Programs freely and instruct Temporal Server to send new 
+executions to the new Build ID and let old executions run on their old Build IDs to
+completion.
+
+When using Worker Versioning, you may need to run multiple versions of your Worker for
+some time, until Workers of old Build IDs are not needed anymore (i.e. they are not
+reachable by any current or future Workflow Execution). Temporal provides Reachability
+API to help determine when a Worker version can be decommissioned. 
+
+## When and why you should use Worker Versioning
+
+The main reason to use this feature is that it frees you from having to worry about
+nondeterministic changes. This is a significant gain, however it comes at the cost of
+running multiple versions of Workers simultaneously.
+
+For this reason, Worker Versioning is best suited for short-lived Workflows. Because 
+the old Build ID is only needed to be kept for a short window during/after a deployment
+until all open workflows belonging to it close.
+
+For long-running Workflows, you would need to keep multiple versions of your Workers 
+running for an extended duration, or avoid making any nondeterministic changes by using
+[Patching](https://docs.temporal.io/workflows#patching) or other means.
+
+There are also ways to break some long-running Workflows to shorter ones to benefit from
+Worker Versioning.
+
+## Worker Versioning Rules
+
+Temporal Server allows you to manage routing of tasks to Build IDs via 
+**Worker Versioning Rules**.
+
+Worker Versioning rules and added to a given Task Queue. If your Worker Program contains 
+multiple Workers polling multiple Task Queues, you would need to separately update the rules 
+of each Task Queue.
+
+There are two types of rules: _Build ID Assignment rules_ and _Compatible Build ID Redirect rules_.
```

Comment:
could we just call them "Build ID Redirect rules"? they must be compatible, yes, but maybe it doesn't have to be part of the official name?